rust/kernel/str.rs

```diff
@@ -500,3 +500,77 @@ impl fmt::Write for Formatter {
         }
     }
 }
+
+/// An owned string that is guaranteed to have exactly one `NUL` byte, which is at the end.
+///
+/// Used for interoperability with kernel APIs that take C strings.
+///
+/// # Examples
+///
+/// ```
+/// # use kernel::prelude::*;
+/// use kernel::str::CString;
+///
+/// let s = CString::try_from_fmt(fmt!("{}{}{}", "abc", 10, 20)).unwrap();
+/// assert_eq!(s.as_bytes_with_nul(), "abc1020\0".as_bytes());
+///
+/// let tmp = "testing";
+/// let s = CString::try_from_fmt(fmt!("{tmp}{}", 123)).unwrap();
+/// assert_eq!(s.as_bytes_with_nul(), "testing123\0".as_bytes());
+/// ```
+pub struct CString {
+    buf: Vec<u8>,
+}
```

Comment:
Could you add a ToOwned impl for CStr with CString as owned variant? That allows Cow<CStr> to be used. Also having From<CStr> and From<CString> impls for Cow<Cstr> would be nice for convenience.

Reply:
It would indeed be nice to have, but alas I don't think this is usable in kernel code because there is an allocation involved that could fail.

So we'd really need something like `TryToOwned` with a fallible `try_to_owned` function. Defining this is outside the scope of this PR. I opened #678 to keep track of this.